### PR TITLE
RFR: Env variables support paramiko

### PR DESCRIPTION
**Pull request by lakshmi-kannan**
Spent several hours on this and tested various cases. Seems to work now. But would appreciate more eyes and testing.  cc: @manasdk , @Kami

### CODE REVIEW

File: st2common/st2common/models/system/action.py [hunk 1]
```diff
@@ -86,6 +88,18 @@ def get_timeout(self):
     def get_cwd(self):
         return self.cwd
 
+    def _get_env_vars_export_string(self):
```
Comment (lakshmi-kannan) [inline st2common/st2common/models/system/action.py]: Note to self: needs unit test

**Comment by lakshmi-kannan** [comment]
Also found one more issue where we do cd {{cwd_dir}} && sudo -E -- bash -c 'command'. This is broken if only sudo user has access to cwd. So I'll be adding a commit to fix that.

**Comment by lakshmi-kannan** [comment]
This still doesn't get this command to work on a box: https://github.com/StackStorm/st2cd/blob/master/actions/workflows/st2_deploy_test.yaml#L42

I am debugging that.

**Comment by Kami** [comment]
(thinking out loud)

1. Does paramiko indeed use `export` or does it just use local variables and `FOO=bar ./script.sh` syntax? export is probably better in this case, especially if user / we run multiple commands...
2. export only works in some shells (sh, bash) but not others (csh), but I don't think that's a big deal since we pretty much assume bash.

**Comment by Kami** [comment]
> Also found one more issue where we do cd {{cwd_dir}} && sudo -E -- bash -c 'command'. This is broken if only sudo user has access to cwd. So I'll be adding a commit to fix that.

Good catch. We should indeed cwd in the sudo block.

**Comment by lakshmi-kannan** [comment]
>Does paramiko indeed use export or does it just use local variables and FOO=bar ./script.sh syntax? export is probably better in this case, especially if user / we run multiple commands

I manually create an ``export`` string and add it to the command. E.g. ``export FOO=BAR && /tmp/exec_id/script.py --arg1=foo``

>export only works in some shells (sh, bash) but not others (csh), but I don't think that's a big deal since we pretty much assume bash.
Yep. This is what fabric does too. Too bad if they use csh :P.

**Comment by lakshmi-kannan** [comment]
I'd definitely appreciate :eyes: on the tests. I think now we have better coverage than before. There are still more tests I could write. If there is anything in particular you want me to add, let me know. This PR was a PITA.

File: st2common/st2common/models/system/action.py [hunk 2]
```diff
@@ -86,6 +88,18 @@ def get_timeout(self):
...
+        if self.env_vars:
+            exports = ' '.join(
+                '%s=%s' % (quote_unix(k), quote_unix(v))
```
Comment (manasdk) [inline st2common/st2common/models/system/action.py]: AFAICT quote_unix will run on exports later when the complete command is prepared. Is it still necessary to quote here?

Comment (lakshmi-kannan) [inline st2common/st2common/models/system/action.py]: Trust me, it's needed :P.

File: st2common/tests/unit/test_paramiko_script_action_model.py [hunk 3]
```diff
@@ -20,7 +20,107 @@
...
 
-    def test_get_command_string(self):
+    def test_get_command_string_no_env_vars(self):
+        script_action = ParamikoRemoteScriptActionTests._get_test_script_action()
+        ex = 'cd /tmp && /tmp/remote_script.sh song=\'b s\' \'taylor swift\''
+        self.assertEqual(script_action.get_full_command_string(), ex)
+
+        # Test with sudo
+        script_action.sudo = True
+        ex = 'sudo -E -- bash -c ' + \
```
Comment (manasdk) [inline st2common/tests/unit/test_paramiko_script_action_model.py]: lol ... I look at this an accept it as fact

File: st2common/st2common/models/system/action.py (hunk 2 above)
Comment (manasdk) [inline st2common/st2common/models/system/action.py]: Then please add a comment saying why it is needed. Else, it could easily be dropped in some refactoring/clean-up effort.

**Comment by manasdk** [comment]
Seems like a liberal sprinkling of quote_unix  fixes our woes. For me very hard to rationalize this PR.

So +1

**Comment by Kami** [comment]
> Seems like a liberal sprinkling of quote_unix fixes our woes. For me very hard to rationalize this PR.

The problem with "liberally sprinkling" quote_unix everywhere is that things will break if values are escaped multiple times.

As mentioned before, imo, escaping is same as handling unicode / bytes in Python 3 - as long as you do it explicitly and you know in every step you are dealing with (bytes / unicode, unescaped / escape) value, there will be no problems.

Anyway, yeah, the biggest problem is it's kinda hard to follow through those changes right now and know where the value is escaped.

Maybe in the future we should come up with some kind of class abstraction for escaped values which behaves like `str`. It might be a bit heavy and unpythonic, but the idea behind is that we can do "static" analysis / lint and know what we are dealing with in each line.

**Comment by Kami** [comment]
Also, I know that I'm late, but LGTM :+1:
